cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1398,20 +1420,62 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
+    add_default_name(table_meta->column_metadata[i], "_col" + std::to_string(i));
   }
+}
 
-  auto vec = table_to_linked_columns(table);
-  auto schema_tree =
-    construct_schema_tree(vec, *_table_meta, _single_write_mode, _int96_timestamps);
+/**
+ * @brief Perform the processing steps needed to convert the input table into the output Parquet
+ * data for writing, such as compression and encoding.
+ *
+ * @param input The input table
+ * @param partitions Optional partitions to divide the table into, if specified then must be same
+ *        size as number of sinks
+ * @param table_meta The table metadata
+ * @param curr_agg_meta The current aggregate writer metadata
+ * @param kv_meta Optional user metadata
+ * @param max_page_fragment_size_opt Optional maximum number of rows in a page fragment
+ * @param max_row_group_size Maximum row group size, in bytes
+ * @param max_page_size_bytes Maximum uncompressed page size, in bytes
+ * @param max_row_group_rows Maximum row group size, in rows
+ * @param max_page_size_rows Maximum page size, in rows
+ * @param column_index_truncate_length maximum length of min or max values in column index, in bytes
+ * @param stats_granularity Level of statistics requested in output file
+ * @param compression Compression format
+ * @param dict_policy Policy for dictionary use
+ * @param max_dictionary_size Maximum dictionary size, in bytes
+ * @param single_write_mode Flag to indicate that we are guaranteeing a single table write
+ * @param int96_timestamps Flag to indicate if timestamps will be written as INT96
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A tuple of the intermediate results containing the processed data
+ */
+auto convert_table_to_parquet_data(table_view const& input,
+                                   std::vector<partition_info> const& partitions,
+                                   table_input_metadata& table_meta,
+                                   std::unique_ptr<aggregate_writer_metadata> const& curr_agg_meta,
+                                   std::vector<std::map<std::string, std::string>> const& kv_meta,
+                                   std::optional<size_type> max_page_fragment_size_opt,
+                                   size_t max_row_group_size,
+                                   size_t max_page_size_bytes,
+                                   size_type max_row_group_rows,
+                                   size_type max_page_size_rows,
+                                   int32_t column_index_truncate_length,
+                                   statistics_freq stats_granularity,
+                                   Compression compression,
+                                   dictionary_policy dict_policy,
+                                   size_t max_dictionary_size,
+                                   bool single_write_mode,
+                                   bool int96_timestamps,
```

Comment:
I wish we could pass the options object directly here :(
Maybe there's a case for an `encode_options` so we group most of these? Just not in this PR, IMO.